# OpenCMS
 Custom Mass Spectrometry library 
This is a repository containing the code of the OpenCustomDB web-based platform.
The platform uses the OpenCustomDB python package, available at: https://openprot.org/opencustomdb/

## Submitting an analysis to the OpenCustomDB website ## 
To submit an analysis, click on the Run Analysis tab on the menu bar. You should get to this page:

![image](https://user-images.githubusercontent.com/65976013/170891347-5886043a-0057-4b10-b039-da6ddfcd9314.png)


In the first two boxes, you are asked to enter **an email adress** , which will be used to send you a unique link to the results of your analysis; and **a study name**, which will be used to refer to your analysis. Then, you are asked to select the adequate parameters for your analysis. As OpenCustomDB starts either from an **unannotated VCF file** or an annotated VCF in tsv format, these parameters should be the one used to build the inputs file. You will have to select the species you are working with, the genome assembly used (please note the differences between hg38, hg19 and b37) and the genome build you wish to use to annotate your VCF file (common or deep genome annotation).	
Before clicking on "Run analysis", you must upload your mandatory inputs (see section below). A progress bar will appear and a message confirming the upload of your data will be displayed if the upload was completed successfully!
 Once your analysis has been successfully submitted, a blue box will appear at the bottom of your screen. The link in this box is unique to your analysis and will be emailed to you once the analysis is done.
 
 ![image](https://user-images.githubusercontent.com/65976013/170891416-a5bd7015-e1c8-41fa-8be6-56797ca329ea.png)

 

## Mandatory inputs:
**•	Either an unannotated VCF file or an annotated tsv file.** The former can be generated by any variant caller as : SAMtools, the GATK Best Practices pipeline, CTAT, FreeBayes, MuTect2, Strelka2, and VarScan2.
The annotated tsv file can be generated with output of variant annotator as ANNOVAR or SnpEff, see an example below. 
Prot	Transcrit	HGVS_P	HGVS_C	Potential_Error
IP_3379289	FBgn0013687	p.Lys11Asn	c.33A>T	WARNING_TRANSCRIPT_NO_START_CODON
IP_3379291	FBgn0013687	p.Lys21Ile	c.62A>T	
FBpp0099974	FBtr0005088	p.Val575Val	c.1725G>C	
FBpp0099974	FBtr0005088	p.Val575Val	c.1725G>C	
FBpp0099544	FBtr0006151	p.Phe162Phe	c.486T>C	

	To be able to detect variants of non-canonical proteins we recommend the use of an unannotated VCF file as it uses the variant annotator OpenVar https://openprot.org/openvar/.
**•	A transcript expression file in tsv format.** This can be generated with Kallisto and should be presented as shown below.
target_id	length	eff_length	est_counts	tpm
FBtr0332346	69016	68805.1	2.36478	0.000273
FBtr0332348	69435	69224.1	34.6621	0.003977
FBtr0332354	50529	50318.1	21509.1	3.3952

## Parameters:
**•	Proteins Numbers** by default, the 100,000 proteins encoded in the most abundant transcripts in the sample are annotated in the generated database. This value can be changed. Rather than limiting the total number of proteins, TPM values can be used instead.
**•	Alternative proteins** active by default. While deactivated non-canonical protein will not be added in the database.

## Optional inputs:
**•	A list of transcripts to add to the analysis.** Canonical and non-canonical proteins transcribed out of those transcripts will be added even if it goes over the database’s size limits.
**•	A list of transcript to exclude of the analysis.** Canonical and non-canonical proteins transcribed out of those transcripts will be removed

## Understanding the results 
Once the analysis is done, you will receive an email with a link to your results. The results page is organized in boxes, as follows.
Downloading your results
OpenCustomDB output folder contains the following:
**•	A protein database in fasta format.** This is the main output of OpenCustomDB. The database is ready to use with any MS/MS search engine. It contains canonical proteins, altProt (IP_), novel isoforms (II_). Variants of each category are annotated with an @.
**•	A tsv file listing all consequences for each variant.** This file would thus contain as many lines as there are consequences for all of the submitted variants. For example, a SNP altering a canonical ORF and an alternative ORF would have 2 lines, one listing the effect on the canonical ORF, the second listing the effect on the alternative ORF.
**•	 A summary.txt file.** This file gives you indication about the numbers of proteins present of each category as well as the expression quantification in TPM of the last transcript selected.
 
## Command line.

OpenCustomDB is also available in command line.

You can download OpenCustomDB with all the databases needed here : https://api.openprot.org/api/2.0/files/OpenCustomDB/OpenCustomDatabase.zip

Then Follow the instruction below.
**Launch from Python as shown below :**


**Instruction :**

**1**- Go to parent directory of OpenCustomDatabase (home).

**2**- open python terminal
**3**- paste :
``` from OpenCustomDatabase.OpenCustomDB import * ```

**4**- paste : path of your vcf, name of your experiment, path of your kallisto input file.
```
OPCMS = OpenCMS(vcf_path="vcf",
		   expname = "name",
		   input_kallisto= "kallisto",
		   annotation = "human-OP_Ensembl")
```
**5**-paste :
```
OPCMS.run()
```

6-an output folder with the name of the vcf should have been created.

## TestFile
For testing purposes a VCF file and a kallisto output file are available on this github for testing purposes
Test_VCF_OCDB_humanHG38.vcf
kallisto_test_file_human_hg38.tsv
